src/index.tsx

```diff
@@ -13,72 +12,66 @@ interface AccorionProps {
   [rest: string | number | symbol]: unknown;
 }
 
-let interval: number;
+interface ListItemProps {
+  SummaryComponent: React.ElementType;
+  DetailComponent: React.ElementType;
+  id: string | number;
+}
 
-const Accordion = ({ items, ...rest }: AccorionProps) => {
-  const [opened, setOpened] = useState<Record<string, boolean>>({});
+const ListItem = ({
+  id,
+  SummaryComponent,
+  DetailComponent,
+  ...rest
+}: ListItemProps) => {
+  const contentItem = useRef<HTMLDivElement>(null);
+  const [opened, setOpened] = useState<boolean>(false);
 
   const clickHandler = (e: MouseEvent): void => {
-    let element = e.target as HTMLElement;
-
-    if (element.parentElement?.tagName === "LI") {
-      element = element.parentElement;
-    }
-
-    if (element.tagName !== "LI") return;
-
-    const id = element.getAttribute("id");
-
-    if (!id) return;
-
-    const isOpen = !!opened[id];
-
-    if (isOpen) {
-      const contentItem = document.getElementById(`acc-item-${id}`);
-
-      if (!contentItem) return;
-
-      contentItem
-        .animate(
-          { maxHeight: 0, opacity: 0 },
-          { duration: 100, easing: "ease-out" }
-        )
-        .finished.then(() => {
-          setOpened((prv) => ({ ...prv, [id]: false }));
-        });
-      return;
-    }
-
-    setOpened((prv) => ({ ...prv, [id]: true }));
-
-    // listen for DOM to be added
-    interval = setInterval(() => {
-      const contentItem = document.getElementById(`acc-item-${id}`);
-      if (!contentItem) return;
-      if (contentItem?.scrollHeight) {
-        const scrollHeight = contentItem.scrollHeight;
-
-        contentItem.animate(
-          { maxHeight: `${scrollHeight}px`, opacity: 1 },
-          { duration: 100, easing: "ease-in", fill: "forwards" }
-        );
-        clearInterval(interval);
+    if (opened && contentItem.current) {
+      const element = e.target as HTMLElement;
+      const clickedInside = contentItem.current.contains(element);
+      if (!clickedInside) {
+        contentItem.current
+          .animate(
+            { maxHeight: 0, opacity: 0 },
+            { duration: 100, easing: "ease-out" }
+          )
+          .finished.then(() => setOpened(false));
       }
-    }, 5);
+    } else {
+      setOpened(true);
+    }
   };
 
   useEffect(() => {
-    // remove on unmount
-    return clearInterval(interval);
-  }, []);
+    if (opened && contentItem && contentItem.current?.scrollHeight) {
+      const scrollHeight = contentItem.current.scrollHeight;
+      contentItem.current.animate(
+        { maxHeight: `${scrollHeight}px`, opacity: 1 },
+        { duration: 100, easing: "ease-in", fill: "forwards" }
+      );
+    }
+  }, [opened]);
 
   return (
-    <ul onClick={clickHandler}>
-      {items.map(({ id, ...data }) => (
-        <ListItem id={id} {...data} key={id} isOpen={opened[id]} {...rest} />
-      ))}
-    </ul>
+    <li onClick={clickHandler} className="acc-item">
```

Comment:
Don't you think adding a click handler on every element will cause some performance issues?

Reply:
No, I'll post screenshots of the performance flame graph, in my initial tests it looks like this is far more performant.

Adding event listeners by themselves is not a performance issue. Having many events firing could be a performance issue, but in this case you have only one event at a time. The main difference is can be seen just by breaking down what's happening within the app when you click on an accordion.

Currently its:
1. Click event fired
2. Interval function ticking
3. Document.getElementById
4. Element animate()
5. 400 re-renders

With this PR its:
1. Click event fired
2. Element animate()
3. 1 re-render

Reply:
Changes in the PR - https://imgur.com/a/VEtWPyY
Current code - https://imgur.com/a/YLvoivT

There is so much less going on in the first photo - even though every element has an event listener, and every element will update their state, it only does it one at a time. Every click is one event, one state update, one re-render. That's why it's so much more performant.

Reply:
I am not sure if you just use state individually, how will you close other accordions?
I've added it as new feature.

Thanks for all the effort you put

Reply:
No problem!

As for closing other Accordions, what I'd probably do is something to share the "open/close" logic, maybe as a hook, so when one item opens, all the other items are aware and automatically close. 

For example, convert that logic to a `const [opened, setOpened] = useToggleState(false)`

and inside the useToggleState, do what you did initially, make the `opened` state the `id` of the element that is open.

Then your `isOpen` will just be `open === element.id`. 

This could also be done by firing an event, but that's not ideal.

Reply:
Also, just wanted to mention, using a MutationObserver will have the same performance issues that using the interval - it's slightly better since you're removing the setInterval, but the MutationObserver will still include the hundreds of re-renders per click.